apps/rush-lib/src/logic/DependencyIntegrator.ts

```diff
@@ -0,0 +1,355 @@
+      throw new Error(`Cannot auto-detect dependency type of "${packageName}" for project "${project.packageName}"`);
+    }
+
+    if (!dependencyKind) {
+      dependencyKind = oldDependencyKind;
+    }
+
+    // update the dependency
+    if (dependencyKind === DependencyKind.Dependency) {
+      packageJson.dependencies = this._updateDependency(packageJson.dependencies, packageName, newVersion);
+    } else if (dependencyKind === DependencyKind.DevDependency) {
+      packageJson.devDependencies = this._updateDependency(packageJson.devDependencies, packageName, newVersion);
+    }
+
+    if (!doNotSave) {
+      // overwrite existing file
+      const packageJsonPath: string
+        = path.join(project.projectFolder, FileConstants.PackageJson);
+      JsonFile.save(project.packageJson, packageJsonPath);
+
+      console.log(colors.green('Wrote ') + packageJsonPath);
+    }
+  }
+
+  private _getNormalizedVersionSpec(
+    packageName: string,
+    initialSpec: string | undefined,
+    implicitlyPinnedVersion: string | undefined,
+    rangeStyle: SemVerStyle): string {
+
+    console.log(colors.gray(`Determining new version for dependency: ${packageName}`));
+    if (initialSpec) {
+      console.log(`Specified version selector: ${colors.magenta(initialSpec)}`);
+    } else {
+      console.log(`No version selector specified, will be automatically determined.`);
+    }
+    console.log();
+
+    // if ensureConsistentVersions => reuse the pinned version
+    // else, query the registry and use the latest that satisfies semver spec
+    if (initialSpec && implicitlyPinnedVersion && initialSpec === implicitlyPinnedVersion) {
+      console.log(colors.green('The specified version ')
+        + colors.magenta(initialSpec)
+        + colors.green(' has been selected as it matches the implicitly preferred version.'));
+      return initialSpec;
+    }
+
+    if (this._rushConfiguration.ensureConsistentVersions && !initialSpec && implicitlyPinnedVersion) {
+      console.log(colors.grey('The enforceConsistentVersions policy is currently active.'));
+      console.log(`Using the implicitly preferred version ${colors.magenta(implicitlyPinnedVersion)}`);
+      return implicitlyPinnedVersion;
+    }
+
+    let selectedVersion: string | undefined;
+
+    if (initialSpec && initialSpec !== 'latest') {
+      console.log(colors.gray('Finding newest version that satisfies the selector: ') + initialSpec);
+      console.log();
+      console.log(`Querying registry for all versions of ${packageName}...`);
+      const allVersions: string =
+        Utilities.executeCommandAndCaptureOutput(this._rushConfiguration.packageManagerToolFilename,
+          ['view', packageName, 'versions', '--json'],
+          this._rushConfiguration.commonTempFolder);
+
+      let versionList: Array<string> = JSON.parse(allVersions);
+      versionList = versionList.sort((a: string, b: string) => { return semver.gt(a, b) ? -1 : 1; });
```

Comment:
>semver.gt [](start = 72, length = 9)

Is this sorting operation expensive?

Reply:
Yeah but we need it

---
In reply to: [220767412](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220767412) [](ancestors = 220767412)